src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs

```diff
@@ -44,6 +45,7 @@ internal abstract class SourceOrdinaryMethodSymbolBase : SourceOrdinaryMethodOrU
                  isIterator: isIterator)
         {
             Debug.Assert(diagnostics.DiagnosticBag is object);
+            Debug.Assert(!isReadOnly || this.IsImplicitlyDeclared, "We only expect synthesized methods to use this flag to make a method readonly. Explicitly declared methods should get this value from modifiers in syntax.");
```

Comment:
Consider also asserting that we ever only set `isReadOnly` inside a struct `containingType`.